tests/test_api.py

```diff
@@ -786,3 +786,160 @@ def test_product_components(client, api_path):
+        └──(PROVIDES) dep1_comp
+
+    where upstream is defined as a child node of a root comp and dep1 is also defined as a root component
+    /
+    └──(SOURCE) dep1_comp: top level root component
+            └──(PROVIDES) dep2_comp
+
+    what are the expectations in terms of sources, providers and upstreams on each of these components ?
+
+    +--------------------+---------+----------+-----------+
+    |                    | sources | provides | upstreams |
+    +--------------------+---------+----------+-----------+
+    | root_component     | 0       | 2        | 1         |
+    +--------------------+---------+----------+-----------+
+    | upstream_component | 0       | 0        | 0         |
+    +--------------------+---------+----------+-----------+
+    | dep1_component     | 1       | 1        | 0         |
+    +--------------------+---------+----------+-----------+
+    | dep2_component     | 2       | 0        | 0         |
+    +--------------------+---------+----------+-----------+
+
+    Are the following statements true ?
+
+    * upstream has 0 sources as it is its own SOURCE
+    * should dep1 or dep2 dynamically resolve the top level component's upstreams ?
+    * should dep2 list 2 sources ?
+    * we should not list upstreams in provides right ?
+
+    """
+
+    # create a top level root source component
+    root_comp = ComponentFactory(
+        name="root_comp",
+        type=Component.Type.RPM,
+        arch="src",
+        namespace=Component.Namespace.REDHAT,
+        related_url="https://example.org/related",
+    )
+    root_node, _ = ComponentNode.objects.get_or_create(
+        type=ComponentNode.ComponentNodeType.SOURCE,
+        parent=None,
+        purl=root_comp.purl,
+        defaults={"obj": root_comp},
+    )
+
+    # create upstream child node
+    upstream_comp = ComponentFactory(
+        name="upstream_comp",
+        type=Component.Type.PYPI,
+        namespace=Component.Namespace.UPSTREAM,
+        related_url="https://example.org/related",
+    )
+    ComponentNode.objects.get_or_create(
+        type=ComponentNode.ComponentNodeType.SOURCE,
+        parent=root_node,
+        purl=upstream_comp.purl,
+        defaults={"obj": upstream_comp},
+    )
+
+    # create dep component
+    dep_comp = ComponentFactory(
+        name="cool_dep_component",
+        arch="src",
+        type=Component.Type.RPM,
+        namespace=Component.Namespace.REDHAT,
+    )
+    # make it a child of root component
+    ComponentNode.objects.get_or_create(
+        type=ComponentNode.ComponentNodeType.PROVIDES,
+        parent=root_node,
+        purl=dep_comp.purl,
+        defaults={"obj": dep_comp},
+    )
+    # also make it its own top level root component
+    dep_node, _ = ComponentNode.objects.get_or_create(
+        type=ComponentNode.ComponentNodeType.SOURCE,
+        parent=None,
+        purl=dep_comp.purl,
+        defaults={"obj": dep_comp},
+    )
+
+    # create 2nd level dep
```

Comment:
good idea will add this as a separate test

Reply:
this test is not specifically about adding coverage - it is about having a single standalone test to ensure we all share common understanding ... I will make a 2nd test with OCI example

Reply:
Agree we should probably just rename the components to match the Container structure.

Although a second test which outlines the RPM structure may also be informative. It would be the same as James provides here, just without the 2nd level of provided dependency.

Reply:
added 2nd test